clients/src/test/java/org/apache/kafka/common/metrics/SensorTest.java

```diff
@@ -124,4 +134,48 @@ public void testIdempotentAdd() {
         assertEquals(1, sensor.metrics().size());
         assertEquals(org.apache.kafka.common.metrics.stats.Avg.class, sensor.metrics().get(0).measurable().getClass());
     }
+
+    /**
+     * The Sensor#checkQuotas should be thread-safe since the method may be used by many ReplicaFetcherThreads.
+     */
+    @Test
+    public void testCheckQuotasInMultiThreads() throws InterruptedException, ExecutionException {
+        final Metrics metrics = new Metrics(new MetricConfig().quota(Quota.upperBound(Double.MAX_VALUE))
+            // decreasing the value of time window make SampledStat always record the given value
+            .timeWindow(1, TimeUnit.MILLISECONDS)
+            // increasing the value of samples make SampledStat store more samples
+            .samples(100));
+        final Sensor sensor = metrics.sensor("sensor");
+
+        assertTrue(sensor.add(metrics.metricName("test-metric", "test-group"), new Rate()));
+        final int threadCount = 10;
+        final CountDownLatch latch = new CountDownLatch(1);
+        ExecutorService service = Executors.newFixedThreadPool(threadCount);
+        List<Future<Throwable>> workers = new ArrayList<>(threadCount);
+        for (int i = 0; i != threadCount; ++i) {
+            final int index = i;
+            workers.add(service.submit(new Callable<Throwable>() {
+                @Override
+                public Throwable call() {
+                    try {
+                        assertTrue(latch.await(5, TimeUnit.SECONDS));
+                        for (int j = 0; j != 20; ++j) {
+                            sensor.record(j * index, System.currentTimeMillis() + j, false);
+                            sensor.checkQuotas();
+                        }
+                        return null;
+                    } catch (Throwable e) {
+                        return e;
+                    }
+                }
+            }));
+        }
+        latch.countDown();
+        service.shutdown();
```

Comment:
we need to shutdown the executor service even if there is a test failure (either in a finally block or the teardown).